File: BinarySearch/Java/OnAnswers/KokoBananas.java
public class KokoBananas {
    public int minimumRateToEatBananas(int[] nums, int h) {
        int n = nums.length, maxi = Integer.MIN_VALUE;

        for(int i:nums){
            maxi = Math.max(i,maxi);
        }

        int low = 1, high = maxi;

        while(low<=high){
            int mid = (low+high)/2;
            if(calculateTotalHours(nums,mid)<=h){
                high = mid - 1;
            }
            else {
                low = mid + 1;
            }
        }

        return low;
    }

    private static long calculateTotalHours(int[] nums, int hourly) {
        long totalHours = 0;
        for(int i:nums){
            totalHours += Math.ceil((double)(i)/(double)(hourly));
        }

        return totalHours;
    }

    public static void main(String[] args) {
        KokoBananas sol = new KokoBananas();
        int[] nums = {3,6,7,11};
        int h = 8;
        int result = sol.minimumRateToEatBananas(nums, h);
        System.out.println("The minimum eating rate to finish bananas in " + h + " hours is: " + result);
    }
}

File: BinarySearch/kokoBananas.py
import math
def kokoBananas(piles,h):
    # Intution is to find the maximum element of the array
    # And the range is b/n 1->max(arr). Apply binary search
    # and the result is when the low reaches the pivot point
    low, high = 1,max(piles)
    ans = 1
    while low<=high:
        mid = (low+high)//2
        # Find time to eat each pile
        totalhours = getHours(piles,mid)
        # If the total hours accepts the mid value
        if totalhours<=h:
            res = mid # Possible value but you still go left
            high = mid - 1 # Eliminate the right half
        # If the total hours is less than the target
        else:
            low = mid + 1 # Eliminate the left half
    
    return res # can be 'low' as it is the pivot

def getHours(piles,bananasperhour):
    totalhrs = 0
    # Pass across the piles and calculate time to eat each pile
    for i in range(len(piles)):
        totalhrs += math.ceil(piles[i] / bananasperhour)
    return totalhrs


def main():
    piles = [30,11,23,4,20]
    h = 5
    res = kokoBananas(piles,h)
    print("The minimum hours to eat the bananas are: ",res)


if __name__ == "__main__":
    main()